### PR TITLE
German Credit Dataset Support

### PR DESCRIPTION
**Pull request by eladprager**
update GermanCredit dataset, preprocess, and model
update preprocessor to fit recent master changes

### CODE REVIEW

File: src/pipeline/BankMarketingProductionModel.sav [hunk 1]
```diff
@@ -0,0 +1 @@
+�}�.
```
Comment (guyfreund) [inline src/pipeline/BankMarketingProductionModel.sav:1]: Please delete, this moved to `src/pipeline/model/raw_files/BankMarketingProductionModel.sav`
You may need to retrain

File: src/pipeline/datasets/training_datasets.py [hunk 2]
```diff
@@ -31,4 +33,9 @@ def __init__(self):
         )
 
     def load(self) -> pd.DataFrame:
-        raise NotImplementedError  # TODO: implement
+        names = ['existingchecking', 'duration', 'credithistory', 'purpose', 'creditamount',
+                 'savings', 'employmentsince', 'installmentrate', 'statussex', 'otherdebtors',
+                 'residencesince', 'property', 'age', 'otherinstallmentplans', 'housing',
+                 'existingcredits', 'job', 'peopleliable', 'telephone', 'foreignworker', 'classification']
```
Comment (guyfreund) [inline src/pipeline/datasets/training_datasets.py:39]: what is names?

File: src/pipeline/datasets/deployment_datasets.py [hunk 3]
```diff
@@ -48,7 +49,7 @@ def __init__(self):
         )
 
     def load(self) -> pd.DataFrame:
-        raise NotImplementedError  # TODO: implement
+        return pd.read_csv(self._path, names=GermanCreditDataset.names, delimiter=' ')
```
Comment (guyfreund) [inline src/pipeline/datasets/deployment_datasets.py:52]: what is names?

File: src/pipeline/datasets/deployment_datasets.py [hunk 4]
```diff
@@ -63,4 +64,4 @@ def __init__(self):
...
```
Comment (guyfreund) [inline src/pipeline/datasets/deployment_datasets.py:67]: what is names?

File: src/pipeline/model/paths.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import os
 
 
-BANK_MARKETING_PRODUCTION_MODEL_PATH = os.path.abspath(os.path.join(__file__, "..", "..", "model", "raw_files", "BankMarketingProductionModel.sav"))
\ No newline at end of file
+BANK_MARKETING_PRODUCTION_MODEL_PATH = os.path.abspath(os.path.join(__file__, "..", "..", "model", "raw_files",
+                                                                    "raw_files/BankMarketingProductionModel.sav"))
```
Comment (guyfreund) [inline src/pipeline/model/paths.py:5]: why the change?

File: src/pipeline/model/production_models.py [hunk 6]
```diff
@@ -9,6 +10,7 @@
 from src.pipeline.model.interfaces.imodel import IModel
 from src.pipeline.model.interfaces.imodel_metric import IModelMetric
 from src.pipeline.model.model_metrics import Accuracy, Precision, Recall, F1, AUC
+from sklearn.ensemble import RandomForestClassifier
```
Comment (guyfreund) [inline src/pipeline/model/production_models.py:13]: maybe need to use [https://imbalanced-learn.org/stable/references/generated/imblearn.ensemble.BalancedRandomForestClassifier.html](https://imbalanced-learn.org/stable/references/generated/imblearn.ensemble.BalancedRandomForestClassifier.html)

File: src/pipeline/preprocessing/preprocessor.py [hunk 7]
```diff
@@ -26,15 +29,23 @@ def __init__(self):
 
     def preprocess(self, dataset: Dataset) -> Tuple[pd.DataFrame, pd.DataFrame, List[IFeatureMetrics]]:
 
+        logging.info(f'Dataset Info: num of total features: {len(dataset.categorical_feature_names) + len(dataset.numeric_feature_names)} | '
```
Comment (guyfreund) [inline src/pipeline/preprocessing/preprocessor.py:32]: check for the label issue and everything with Danielle

File: src/pipeline/preprocessing/preprocessor.py [hunk 8]
```diff
@@ -26,15 +29,23 @@ def __init__(self):
...
 
-        self._processed_df = pd.get_dummies(dataset.raw_df, columns=['job', 'marital',
-                                                                     'education', 'default',
-                                                                     'housing', 'loan',
-                                                                     'contact', 'month',
-                                                                     'poutcome'])
-        self._processed_df.y.replace(('yes', 'no'), (1, 0), inplace=True)
+        pd.DataFrame(StandardScaler().fit_transform(dataset.raw_df[dataset.numeric_feature_names]))
+        d = defaultdict(LabelEncoder)
+        dataset.raw_df[dataset.categorical_feature_names].apply(lambda x: d[x.name].fit_transform(x))
```
Comment (guyfreund) [inline src/pipeline/preprocessing/preprocessor.py:41]: why edit the dataset raw df? let's not change the dataset itself.

File: src/pipeline/preprocessing/preprocessor.py [hunk 9]
```diff
@@ -76,11 +87,14 @@ def _save_data_as_pickle(self, dataset_class_name: str):
...
-        X_train, X_test, y_train, y_test = train_test_split(data_X, data_y, test_size=0.3)
-        X_test, X_validation, y_test, y_validation = train_test_split(X_test, y_test, test_size=0.5)
+        logging.info(f'Save Data: {dataset_class_name}.pickle, {dataset_class_name}Plus.pickle, '
+                     f'{dataset_class_name}_FeatureMetricsList.pickle files has been saved')
+
+    def split(self, processed_df: pd.DataFrame, label_column_name: str) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+        data_y = processed_df[label_column_name]
+        data_X = processed_df.drop(label_column_name, axis=1)
+        X_train, X_test, y_train, y_test = train_test_split(data_X, data_y, test_size=Config().preprocessing.split.train_test_split_size)
+        X_test, X_validation, y_test, y_validation = train_test_split(X_test, y_test, test_size=Config().preprocessing.split.validation_test_split_size)
```
Comment (guyfreund) [inline src/pipeline/preprocessing/preprocessor.py:97]: nice

**Review by guyfreund — CHANGES_REQUESTED** [review]
Very Nice!!!

File: src/pipeline/model/production_models.py (hunk 6 above)
Comment (eladprager) [reply src/pipeline/model/production_models.py:13]: I've used their's model import (search for Random Forest in https://www.kaggle.com/hendraherviawan/predicting-german-credit-default)
